Implement NewsSideBar component

EVAST9919 opened the pull request:
Basic implementation with no integration. Integration will come in a separate pr.
Implemented API components have been tested, results on the screenshots:
web:
![pr_web](https://user-images.githubusercontent.com/22874522/117620616-50d18680-b179-11eb-9697-b16e44812d93.png)
pr:
![pr_osu](https://user-images.githubusercontent.com/22874522/117620638-562ed100-b179-11eb-8d74-9ea20ea496b2.png)

EVAST9919 committed:
Implement basic years panel

EVAST9919 committed:
Implement MonthPanel component

EVAST9919 committed:
Implement NewsSideBar component

EVAST9919 committed:
CI fixes

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Remove overcomplicated date logic in MonthPanel

bdach commented on osu.Game/Overlays/News/Sidebar/MonthPanel.cs:
```diff
@@ -0,0 +1,152 @@
+            }, true);
+
+            // First state change should be instant.
+            FinishTransforms();
+            postsFlow.FinishTransforms();
```
```suggestion
            FinishTransforms(true);
```

frenzibyte replied:
Also this won't work on the autosize transforms of `MonthPanel`, they're added way later on at the end of `CompositeDrawable.UpdateSubTree`, there may be a solution for this case somewhere around the game, but the only way I can think is overriding `UpdateAfterAutoSize` and call once instead, or something around that.

bdach commented on osu.Game/Overlays/News/Sidebar/NewsSideBar.cs:
```diff
@@ -0,0 +1,113 @@
+                    return;
+                }
+
+                yearsPanel.Years = metadata.NewValue.Years;
```
not sure how to feel about this type of data propagation. perhaps it might be nicer to cache `Metadata` and have the child components resolve it and propagate changes that way.

bdach commented on osu.Game/Overlays/News/Sidebar/NewsSideBar.cs:
```diff
@@ -0,0 +1,113 @@
+                    {
+                        var month = p.PublishedAt.Month;
+
+                        if (dict.ContainsKey(month))
+                            dict[month].Add(p);
+                        else
+                        {
+                            dict.Add(month, new List<APINewsPost>(new[] { p }));
+                        }
```
can be shortened using linq `.ToLookup(post => post.PublishedAt.Month)`

frenzibyte replied:
was gonna suggest that, but then `post.PublishedAt.Year` comes into play, maybe a tuple could work if it supports equality properly.

bdach replied:
How are years relevant here? The articles shown should be from one year, as per web, no?

frenzibyte replied:
Right, been confused about how this works.

bdach commented on osu.Game/Overlays/News/Sidebar/NewsSideBar.cs:
```diff
@@ -0,0 +1,113 @@
+
+                    foreach (var keyValuePair in dict)
+                    {
+                        monthsFlow.Add(new MonthPanel(keyValuePair.Value)
+                        {
+                            IsOpen = { Value = isFirst }
+                        });
```
how does this ensure that the dictionary does not mess up month order by swapping them? i think it just doesn't? dictionaries have no inherent order, and expecting enumeration order to be the same as insertion order is in fact invoking undefined behaviour which can be broken at any time if dotnet upstream decides to do so. [see docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?redirectedfrom=MSDN&view=net-5.0#remarks):

> For purposes of enumeration, each item in the dictionary is treated as a `KeyValuePair<TKey,TValue>` structure representing a value and its key. The order in which the items are returned is undefined.

EVAST9919 replied:
I guess I should perform sorting by key then to ensure it's in order?

bdach commented on osu.Game/Overlays/News/Sidebar/YearsPanel.cs:
```diff
@@ -0,0 +1,105 @@
+{
+    public class YearsPanel : CompositeDrawable
+    {
+        public int[] Years
+        {
+            set
+            {
+                years.Clear();
+                years.AddRange(value);
+            }
+        }
+
+        private readonly BindableList<int> years = new BindableList<int>();
```
would like to see this gone. possible direction how that can be achieved has been outlined above (caching/resolving bindables), or just publically exposing the bindable list. if you don't care about the nice value change callbacks you could also ultimately do `Bindable<int[]>`, too, i suppose.

EVAST9919 replied:
Or I just can place everything in the ctor so just property will do. But will cache metadata anyway.

EVAST9919 replied:
Nvm, thare will be no need in property then at all

bdach commented on osu.Game/Overlays/News/Sidebar/MonthPanel.cs:
```diff
@@ -0,0 +1,152 @@
+            {
+                IdleColour = colourProvider.Light2;
+                HoverColour = colourProvider.Light1;
+                Action = () => { }; // TODO
```
as an explanation for this that wasn't given, this is being set because otherwise `ClickableContainer` will spawn in a disabled state, which is a problem because all of these components aren't even supposed to *be* in a disabled state. i guess this *will* go away, though, eventually.

frenzibyte replied:
at least use `Enabled.Value = true` rather than empty TODO comment that just raises questions.

frenzibyte reviewed (changes requested):
Seems fine otherwise.

bdach commented on osu.Game/Overlays/News/Sidebar/MonthPanel.cs:
```diff
@@ -0,0 +1,152 @@
+
+                IsOpen.BindValueChanged(open =>
+                {
+                    icon.Scale = new Vector2(1, open.NewValue ? -1 : 1);
```
I believe flipping is already used in several places. So I see no reason not to keep it.

frenzibyte replied:
If it's a better choice to flip it, then at least use `ChevronUp`, don't really need to do flipping and stuff if the icon exists.

frenzibyte replied:
Yeah disregard that, pointless.

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Add bottom padding for NewsSideBar content

EVAST9919 committed:
Apply suggestions for MonthPanel

EVAST9919 committed:
Use lookup instead of dictionary to distribute posts

EVAST9919 committed:
Cache metadata in NewsSideBar

EVAST9919 committed:
Move metadata change logic to it's own method

EVAST9919 committed:
Rename MonthPanel to MonthDropdown

EVAST9919 commented on osu.Game/Overlays/News/Sidebar/MonthPanel.cs:
```diff
@@ -0,0 +1,152 @@
+
+        public MonthPanel(List<APINewsPost> posts)
+        {
+            Width = 160;
```
Still not sure though. Each tab has a defined size (33.75x15) with 5px spacing between them and the whole container have 5px padding, so total width will be 160px anyway. Hardcoded tab size is easier to maintain because we can just use `FillFlowContainer` instead of using `GridContainer` to make tabs perfectly-centered if we will use relative sizing

EVAST9919 committed:
Add better comments explaining empty actions

EVAST9919 committed:
CI fixes

EVAST9919 committed:
Rename DropdownButton to DropdownHeader

EVAST9919 committed:
Use relative sizing for MonthDropdown

EVAST9919 committed:
Fix TestSceneNewsYearsPanel error

EVAST9919 committed:
Remove pointless test scenes

EVAST9919 committed:
Fix incorrect no posts handling and add corresponding test

EVAST9919 commented on osu.Game/Overlays/News/Sidebar/MonthPanel.cs:
```diff
@@ -0,0 +1,152 @@
+                Spacing = new Vector2(0, 5),
+                Children = new Drawable[]
+                {
+                    new DropdownButton(posts[0].PublishedAt)
```
Not sure we need assert here since I'm pretty sure logic in sidebar will handle everything just right.

frenzibyte replied:
Well, if you have a year and month integers for the month dropdown, and a set of posts that have published dates, it is not exactly known from the context of `MonthDropdown` alone whether all these posts are within the given month and year, and if one post is not within them, then that's basically wrong and `MonthDropdown` has to throw an exception that something is not right with the given data, that's basically one case of what asserts are for in general I believe.

EVAST9919 replied:
If posts are not within a given year - that will be a web-side issue already then?

frenzibyte replied:
In case you missed it, I'm suggesting this assertion when adding separate `year` and `month` integers rather than using the date of the first given post, you can't possibly do any assertion currently.

EVAST9919 replied:
So it will be something like this: MonthPanel(month, year, posts) and checking if post dates are matching these ints?
If that's it - I have no any problems with that

frenzibyte replied:
And use it instead of `posts[0].PublishedAt` in the highlighted line above, which is the main purpose of these integers, yes. 

That's my suggestion from the beginning of this conversation, and the assertion is just for safety measures when the main suggestion itself is applied.

EVAST9919 replied:
Gotcha

EVAST9919 replied:
Done in https://github.com/ppy/osu/pull/12736/commits/f4801c08ff9693c8e97a5252b96a38fe6f288760 commit

bdach commented on osu.Game/Overlays/News/Sidebar/YearsPanel.cs:
```diff
@@ -0,0 +1,105 @@
+        [BackgroundDependencyLoader]
+        private void load(OverlayColourProvider colourProvider)
+        {
+            Width = 160;
```
This size *is* technically wrong. Same deal as with the dropdowns, this isn't ever explicitly sized to 160 pixels. It just comes out that way on chrome.

![2021-05-11-211723_284x133_scrot](https://user-images.githubusercontent.com/20418176/117872053-7a380280-b29e-11eb-9691-b24983a55aeb.png)

But the argument against making the change makes *some* sense. It's not like you couldn't use a grid container but I honestly can't bring myself to care enough about this to belabour the point further.

EVAST9919 committed:
Refactor MonthDropdown to ensure all the posts are within a given month

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Use GridContainer to distribute buttons in YearsPanel

EVAST9919 commented on osu.Game/Overlays/News/Sidebar/MonthPanel.cs:
```diff
@@ -0,0 +1,152 @@
+                    this.ResizeHeightTo(15, 250, Easing.OutQuint);
+
+                    postsFlow.FadeOut(250, Easing.OutQuint);
+                }
```
You forgot about spacing between header and posts. Resizing posts to 0 will not remove spacing, instead there will be a small jump since spacing will be ignored after posts alpha will become 0, Using margin/padding for posts will not fix the issue. Using bottom margin for dropdown will not fix the issue as well since after posts becoming 0 height total dropdown height will be header height + header margin.

frenzibyte replied:
You can add top margin for the posts fill flow container, and wrap it up in a container and perform all the autosizing and resizing logic on it, now the margin would be part of the autosizing.

Also you'll need `AlwaysPresent` on the container that has the autosizing workaround for `UpdateAfterAutoSize` to be invoked properly.

EVAST9919 replied:
Refactored in https://github.com/ppy/osu/pull/12736/commits/315a2b8314a044689eb607dfdced7bc24fc1f4c9

frenzibyte commented on osu.Game/Overlays/News/Sidebar/YearsPanel.cs:
```diff
@@ -0,0 +1,165 @@
+            private Dimension[] getColumnDimensions()
+            {
+                var columnDimensions = new Dimension[column_count];
+                for (int i = 0; i < column_count; i++)
+                    columnDimensions[i] = new Dimension(GridSizeMode.Relative, size: 1f / column_count);
+
+                return columnDimensions;
+            }
+
+            private Drawable[][] createContent()
+            {
+                var buttons = new Drawable[rowCount][];
+
+                for (int i = 0; i < rowCount; i++)
+                {
+                    buttons[i] = new Drawable[column_count];
+
+                    for (int j = 0; j < column_count; j++)
+                    {
+                        var index = i * column_count + j;
+                        buttons[i][j] = index >= years.Length
+                            ? Empty()
+                            : new Container
+                            {
+                                RelativeSizeAxes = Axes.X,
+                                AutoSizeAxes = Axes.Y,
+                                Padding = new MarginPadding
+                                {
+                                    Top = i == 0 ? 0 : spacing / 2,
+                                    Bottom = i == rowCount - 1 ? 0 : spacing / 2,
+                                    Left = j == 0 ? 0 : spacing / 2,
+                                    Right = j == column_count - 1 ? 0 : spacing / 2
+                                },
+                                Child = new YearButton(years[index])
+                            };
+                    }
+                }
+
+                return buttons;
+            }
```
FWIW this could be done in a *slightly* cleaner way by attaching empty cells between each year instead:

<details>
<summary>changes</summary>

```diff
diff --git a/osu.Game/Overlays/News/Sidebar/YearsPanel.cs b/osu.Game/Overlays/News/Sidebar/YearsPanel.cs
index 2a4e2024d2..3c24e580b8 100644
--- a/osu.Game/Overlays/News/Sidebar/YearsPanel.cs
+++ b/osu.Game/Overlays/News/Sidebar/YearsPanel.cs
@@ -96,6 +96,9 @@ private class YearsGridContainer : GridContainer
             private const int column_count = 4;
             private const float spacing = 5f;

+            private Dimension[] rowDimensions;
+            private Dimension[] columnDimensions;
+
             private readonly int rowCount;
             private readonly int[] years;

@@ -106,55 +109,67 @@ public YearsGridContainer(int[] years)

                 RelativeSizeAxes = Axes.X;
                 AutoSizeAxes = Axes.Y;
-                RowDimensions = getRowDimensions();
-                ColumnDimensions = getColumnDimensions();
+                RowDimensions = rowDimensions = getRowDimensions();
+                ColumnDimensions = columnDimensions = getColumnDimensions();
                 Content = createContent();
             }

             private Dimension[] getRowDimensions()
             {
-                var rowDimensions = new Dimension[rowCount];
+                var dimensions = new List<Dimension>();
+
                 for (int i = 0; i < rowCount; i++)
-                    rowDimensions[i] = new Dimension(GridSizeMode.AutoSize);
+                {
+                    dimensions.Add(new Dimension(GridSizeMode.AutoSize));

-                return rowDimensions;
+                    if (i != rowCount - 1)
+                        dimensions.Add(new Dimension(GridSizeMode.Absolute, spacing));
+                }
+
+                return dimensions.ToArray();
             }

             private Dimension[] getColumnDimensions()
             {
-                var columnDimensions = new Dimension[column_count];
+                var dimensions = new List<Dimension>();
+
                 for (int i = 0; i < column_count; i++)
-                    columnDimensions[i] = new Dimension(GridSizeMode.Relative, size: 1f / column_count);
+                {
+                    dimensions.Add(new Dimension(GridSizeMode.Distributed));

-                return columnDimensions;
+                    if (i != column_count - 1)
+                        dimensions.Add(new Dimension(GridSizeMode.Absolute, spacing));
+                }
+
+                return dimensions.ToArray();
             }

             private Drawable[][] createContent()
             {
-                var buttons = new Drawable[rowCount][];
+                var buttons = new Drawable[rowDimensions.Length][];

-                for (int i = 0; i < rowCount; i++)
-                {
-                    buttons[i] = new Drawable[column_count];
+                int y = 0;

-                    for (int j = 0; j < column_count; j++)
+                for (int i = 0; i < rowDimensions.Length; i++)
+                {
+                    if (rowDimensions[i].Mode == GridSizeMode.Absolute)
+                        buttons[i] = new[] { Empty() };
+                    else
                     {
-                        var index = i * column_count + j;
-                        buttons[i][j] = index >= years.Length
-                            ? Empty()
-                            : new Container
+                        buttons[i] = new Drawable[columnDimensions.Length];
+
+                        for (int j = 0; j < columnDimensions.Length; j++)
+                        {
+                            if (columnDimensions[j].Mode == GridSizeMode.Absolute)
+                                buttons[i][j] = Empty();
+                            else
                             {
-                                RelativeSizeAxes = Axes.X,
-                                AutoSizeAxes = Axes.Y,
-                                Padding = new MarginPadding
-                                {
-                                    Top = i == 0 ? 0 : spacing / 2,
-                                    Bottom = i == rowCount - 1 ? 0 : spacing / 2,
-                                    Left = j == 0 ? 0 : spacing / 2,
-                                    Right = j == column_count - 1 ? 0 : spacing / 2
-                                },
-                                Child = new YearButton(years[index])
-                            };
+                                if (y >= years.Length)
+                                    buttons[i][j] = Empty();
+                                else
+                                    buttons[i][j] = new YearButton(years[y++]);
+                            }
+                        }
                     }
                 }
```

</details>

But either way this sounds like something that should be supported framework-side as "grid column / row gaps", later on when more usages of this pop up.

EVAST9919 replied:
To be honest, your logic seems way too hacky (storing dimensions and then comparing their modes), but that's just my preference at this point.

frenzibyte replied:
Hmm, to be fair if this was to be implemented in `GridContainer` it would probably go the way you have it, so yeah disregard that, I think this is good for now.

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Refactor MonthDropdown to decouple autosized logic

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Implement current year highlight in YearsPanel

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Clear transforms in PostsContainer for all children

EVAST9919 committed:
Improve statement readability

frenzibyte commented on osu.Game/Overlays/News/Sidebar/MonthDropdown.cs:
```diff
@@ -0,0 +1,189 @@
+            {
+                RelativeSizeAxes = Axes.X;
+                AutoSizeAxes = Axes.Y;
+                Child = text = new TextFlowContainer(t => t.Font = OsuFont.GetFont(size: 12))
```
```suggestion
                Child = text = new TextFlowContainer(t => t.Font = OsuFont.GetFont(size: 12, weight: FontWeight.Medium))
```

frenzibyte reviewed (commented):
Looks to be matching web, the font sizes seem a bit off though, I'm not sure if it's due to the difference between CSS's font size algorithm and osu!framework's, but on checking, setting a font size of `15px` instead of `12px` makes it almost match osu!web:
| `12px` | `15px` | web (`font-size: 12px`) |
|--------|--------|-----|
| ![CleanShot 2021-05-16 at 09 07 31](https://user-images.githubusercontent.com/22781491/118387278-5a019e00-b626-11eb-9244-fadcd07b17b5.png) | ![CleanShot 2021-05-16 at 09 06 20](https://user-images.githubusercontent.com/22781491/118387277-579f4400-b626-11eb-85e1-81ffe63e98f9.png) | ![CleanShot 2021-05-16 at 09 10 43](https://user-images.githubusercontent.com/22781491/118387330-9e8d3980-b626-11eb-93d4-de5283e55240.png) |

and out of curiosity, this is how web would look with `font-size: 9px`:
![CleanShot 2021-05-16 at 09 22 20](https://user-images.githubusercontent.com/22781491/118387648-47886400-b628-11eb-936d-0f8ac7a109b7.png)


I guess this font size and height mismatching issue persists across all overlays and is known, but is worse on here due to the spacing and the fixed `15px` height on each year button, probably fine for now anyway.

frenzibyte commented:
Actually now I notice this is already tracked in https://github.com/ppy/osu-framework/issues/3271, so this should stay at using `font-size` until it's fixed in the framework instead, the above can be disregarded.

bdach commented:
@frEnzibyte web values are to be replicated 1:1 as long as they're not actively causing a usability problem. this was already stated multiple times on several separate occasions.

frenzibyte commented:
Yeah, I couldn't pay enough attention before but now I realize it more properly.

peppy committed:
SideBar -> Sidebar

peppy committed:
MonthDropdown -> MonthSection

peppy commented:
Is interaction supposed to be working? ie. why does clicking a year not update its selected status? If that relies on the full integration, then at very least this should be emulated locally in the test scene to allow testing the flow.

peppy commented on osu.Game/Overlays/News/Sidebar/YearsPanel.cs:
```diff
@@ -0,0 +1,176 @@
+
+        private class YearButton : OsuHoverContainer
+        {
+            protected override IEnumerable<Drawable> EffectTargets => new[] { text };
```
Unnecessary?

EVAST9919 replied:
There's a slight colour change on hover

peppy replied:
Yes, but check the base implementation. I'm saying this line is unnecessary to keep that effect working.

EVAST9919 replied:
Oh, misunderstood the comment

peppy commented on osu.Game/Overlays/News/Sidebar/YearsPanel.cs:
```diff
@@ -0,0 +1,176 @@
+            {
+                IdleColour = isCurrent ? Color4.White : colourProvider.Light2;
+                HoverColour = isCurrent ? Color4.White : colourProvider.Light1;
+                Action = () => { }; // Avoid button being disabled since there's no proper action assigned.
```
As mentioned in comment, we probably want this to do more than nothing.

EVAST9919 replied:
Relies on integration since we will be calling a method in a NewsOverlay which will trigger a new request, which will reconstruct the whole side bar.

peppy replied:
What I'm saying is that in the test scene this should be hooked up locally. So that it doesn't do nothing.

EVAST9919 replied:
Should be resolved in https://github.com/ppy/osu/pull/12736/commits/586c5c7365b3d0d0291447fb57b316b1e01b098e

peppy commented on osu.Game/Overlays/News/Sidebar/YearsPanel.cs:
```diff
@@ -0,0 +1,176 @@
+
+            public YearsGridContainer(int[] years, int currentYear)
+            {
+                rowCount = (int)Math.Ceiling((float)years.Length / column_count);
```
```suggestion
                rowCount = years.Length / column_count;
```

peppy committed:
Rename content container

peppy committed:
Reorder tests to not have the first test show nothing

peppy committed:
Rewrite `YearsPanel` to not be insanely long

peppy commented on osu.Game/Overlays/News/Sidebar/YearsPanel.cs:
```diff
@@ -0,0 +1,176 @@
+
+namespace osu.Game.Overlays.News.Sidebar
+{
+    public class YearsPanel : CompositeDrawable
```
I started to refactor this to make some sense, but then realised it should be a simple `FillFlow` not a `GridContainer`. I'll leave this to you else this is going to become my own PR.

bdach replied:
Using a grid isn't entirely unwarranted, as web *does* use a 4-column grid layout there, though. However as I said earlier regarding this exact same point, I can't really bring myself to care which one it should be in the end.

peppy replied:
The four column layout can already be forced by setting the size of the columns to relative width 0.25 or similar.

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Simplify YearButton

EVAST9919 committed:
Use FillFlowContainer in YearsPanel

EVAST9919 committed:
Add tooltip and action for PostButton in MonthSection

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Emulate year changes in the test scene

EVAST9919 committed:
Fix filenames does not match contained type

EVAST9919 committed:
Delete TestSceneNewsSideBar.cs

EVAST9919 committed:
Delete NewsSideBar.cs

bdach committed:
Split value change callbacks out to separate methods

bdach committed:
Rename variable to better explain purpose

bdach committed:
Fix weird behaviour in test scene

Due to a callback set up in another place, clicking away from the 2022
year after launching the test scene would remove the 2022 button
(because the callback was returning metadata without it). For simplicity
just trim the 2022 year to make sure both test scenes use the same
consistent set of years.

bdach commented:
have applied further cleanups. seems passable for now.

peppy committed:
Add basic test to show data how one would expect it to be displayed

peppy commented on osu.Game/Overlays/News/Sidebar/MonthSection.cs:
```diff
@@ -0,0 +1,197 @@
+                }
+            }
+
+            private bool autoSizeTransitionApplied;
+
+            // Workaround to allow the dropdown to be opened immediately since FinishTransforms doesn't work for AutoSize{Duration,Easing}.
+            protected override void UpdateAfterAutoSize()
+            {
+                base.UpdateAfterAutoSize();
+
+                if (!autoSizeTransitionApplied)
+                {
+                    AutoSizeDuration = animation_duration;
+                    AutoSizeEasing = Easing.OutQuint;
+
+                    autoSizeTransitionApplied = true;
+                }
+            }
```
I'm not sure what this is trying to achieve, but it looks fine without this logic?

frenzibyte replied:
Makes the posts container autosize immediately on first frame while applying autosize transforms on state changes afterwards.

For some reason this can't be done with a scheduled `FinishTransforms`, although I got to admit I haven't checked thoroughly why, will do in a bit.

bdach replied:
I'm fine with removing. It's not like if this looks bad if the section expand animation plays out on first load instead of instantly.

The reason this can't be done with a schedule is that schedules (both normal, and after children) run too late. [Autosize is the last thing in a subtree update](https://github.com/ppy/osu-framework/blob/5a002b4eeb9598954b26fb86cc93123933d4f52c/osu.Framework/Graphics/Containers/CompositeDrawable.cs#L888).

frenzibyte replied:
I did actually mean a 2-frame scheduling but yeah, what's the point at this point, let the animation play and call it a day.

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into news-sidebar-new

EVAST9919 committed:
Remove convoluted autosize logic in MonthSection

peppy commented on osu.Game/Overlays/News/Sidebar/MonthSection.cs:
```diff
@@ -0,0 +1,180 @@
+                HoverColour = colourProvider.Light1;
+
+                TooltipText = "view in browser";
+                Action = () => host.OpenUrlExternally("https://osu.ppy.sh/home/news/" + post.Slug);
```
Just a heads up that the news system is the only placed these domains are hard-coded. It's going to need to be fixed at some point to use `EndpointConfiguration`.

bdach replied:
Good point, although I expect that action will be going away at some point anyway, for I expect news to be ultimately shown in the client when the implementation is complete.

peppy replied:
Definitely, although I think we will still have "open on web" buttons like in user profile etc.

Anyways that's for another day, as these hard-coded urls are also in other classes this PR doesn't touch.

peppy committed:
Tidy up content and bind event code

peppy committed:
IsOpen -> Expanded
